indexprovider/unsealedstatemanager.go

```diff
@@ -0,0 +1,272 @@
+	if err != nil {
+		return err
+	}
+
+	legacyDeals, err := m.legacyDealsBySectorID(stateUpdates)
+	if err != nil {
+		return fmt.Errorf("getting legacy deals from datastore: %w", err)
+	}
+
+	// For each sector
+	for sectorID, sectorSealState := range stateUpdates {
+		// Get the deals in the sector
+		deals, err := m.dealsBySectorID(ctx, legacyDeals, sectorID)
+		if err != nil {
+			return fmt.Errorf("getting deals for miner %d / sector %d: %w", sectorID.Miner, sectorID.Number, err)
+		}
+
+		// For each deal in the sector
+		for _, deal := range deals {
+			if !deal.AnnounceToIPNI {
+				continue
+			}
+
+			propnd, err := cborutil.AsIpld(&deal.DealProposal)
+			if err != nil {
+				return fmt.Errorf("failed to compute signed deal proposal ipld node: %w", err)
+			}
+			propCid := propnd.Cid()
+
+			if sectorSealState == db.SealStateRemoved {
+				// Announce deals that are no longer unsealed to indexer
+				announceCid, err := m.idxprov.AnnounceBoostDealRemoved(ctx, propCid)
+				if err != nil {
+					// Check if the error is because the deal wasn't previously announced
+					if !errors.Is(err, provider.ErrContextIDNotFound) {
+						// There was some other error, write it to the log
+						usmlog.Errorw("announcing deal removed to index provider",
+							"deal id", deal.DealID, "error", err)
+						continue
+					}
+				} else {
+					usmlog.Infow("announced to index provider that deal has been removed",
+						"deal id", deal.DealID, "sector id", deal.SectorID.Number, "announce cid", announceCid.String())
+				}
+			} else {
+				// Announce deals that have changed seal state to indexer
+				md := metadata.GraphsyncFilecoinV1{
+					PieceCID:      deal.DealProposal.Proposal.PieceCID,
+					FastRetrieval: sectorSealState == db.SealStateUnsealed,
+					VerifiedDeal:  deal.DealProposal.Proposal.VerifiedDeal,
+				}
+				announceCid, err := m.idxprov.announceBoostDealMetadata(ctx, md, propCid)
+				if err != nil {
+					// Check if the error is because the deal was already advertised
+					if !errors.Is(err, provider.ErrAlreadyAdvertised) {
+						// There was some other error, write it to the log
+						usmlog.Errorf("announcing deal %s to index provider: %w", deal.DealID, err)
+						continue
+					}
+				} else {
+					usmlog.Infow("announced deal seal state to index provider",
+						"deal id", deal.DealID, "sector id", deal.SectorID.Number,
+						"seal state", sectorSealState, "announce cid", announceCid.String())
+				}
+			}
+		}
+
+		// Update the sector seal state in the database
+		err = m.sdb.Update(ctx, sectorID, sectorSealState)
+		if err != nil {
+			return fmt.Errorf("updating sectors unseal state in database for miner %d / sector %d: %w", sectorID.Miner, sectorID.Number, err)
+		}
+	}
+
+	return nil
+}
+
+func (m *UnsealedStateManager) getStateUpdates(ctx context.Context) (map[abi.SectorID]db.SealState, error) {
+	// Get the current unsealed state of all sectors from lotus
+	storageList, err := m.api.StorageList(ctx)
```

Comment:
Unfortunately the Lotus api is not picking up deletion of Unsealed copies. In local devnet, I removed an unsealed sector copy and while it displays correctly as removed in the UI (because it does an isUnsealed api check), `StorageList` isn't aware of the change.